### ISSUE TITLE
Provide codegen as sbt sourceGenerator?

### ISSUE THREAD

**Issue by hedefalk** [ghostdogpr/caliban#623]
Sorry to post as "issue" since this might just be a question, but could be a feature request.

I'm trying to understand how to make codegen play nice with sbt. If I understand I need a `Task[Seq[File]]` to stick it into `Compile / sourceGenerators` …?

Looking at the plugin, it delegates to 

https://github.com/ghostdogpr/caliban/blob/4146e8167807901438da65a2b5b3b4dfe7f0e8b5/tools/src/main/scala/caliban/tools/Codegen.scala#L9

but that's a `Task[Unit]`.

The only thing that's important to me is that the codegen is part of the build and that no generated artifacts ever hit SCM. Maybe I'm missing something simple? Seems it would be pretty simple to mod that to `Task[Seq[File]]` though?

**Comment by ghostdogpr** [comment]
That's very possible, the sbt plugin was initially a PR from an external contributor and I have myself very little knowledge of how sbt works. For now I've used the plugin as a one-shot rather than something that runs at every build, but I'm definitely open to making it more ergonomic.

**Comment by hedefalk** [comment]
It seems I'm almost there with an sbt task, just calling the underlying `SchemaWriter.write`. A little bit of copy 'n' paste and hardcoding for my needs right now, but I'll see if it works nicely and I could possibly make a PR if I have a little more time and someone else finds it useful… :)

Thx!

**Comment by ostewart** [comment]
This is a bit of a messy first draft, but in case it helps, here's what's working for me:
```scala
    Compile / sourceGenerators += Def.task {
      val shopifyApiFile = (Compile / sourceManaged).value / "Shopify.scala"
      shopifyApiFile.getParentFile.mkdirs()
      zio.Runtime.default.unsafeRun(
      Codegen.generate(Options.fromArgs(List("shopify-client/schema.gql", shopifyApiFile.getPath, "--packageName", "com.example.shopifyclient")).get, ClientWriter.write)
        .catchAll(reason => zio.console.putStrLn(reason.toString) *> zio.console.putStrLn(reason.getStackTrace.mkString("\n")))
        .as(1)
      )
      Seq(shopifyApiFile)
    }.taskValue
```

**Comment by ghostdogpr** [comment]
Closed by #933
